Simplified Workflow Run Form

jmchilton opened the pull request:
Builds on refactoring the workflow run form landing and orchestration into Vue (#9147).

Implements #9111 - part of https://github.com/orgs/galaxyproject/projects/47 and outlined in http://bit.ly/simplified-workflow-ui.

## What is in the PR:

- Add new Galaxy configuration option - ``simplified_workflow_run_ui`` set to ``off`` by default.
- If instead the admin has set this parameter to ``prefer``, on the workflow run page scan the workflow run request and if three conditions are met show a simplified tool form. These conditions are:
  - No workflow "resource parameters" - pretty niche, not documented, I don't think anyone outside of Canada is using them and even them I'm not sure if they ever got to production.
  - No "open" tools (i.e. tools with disconnected runtime inputs).
  - No "replacement parameters" defined outside PJAs. I'm calling #{} inside PJA and ${} inside tool steps both "replacement parameters" because the user populates them the same way in the GUI. The API only handles them inside the context of the PJA - it seems the GUI is responsible for replacing them at runtime in the traditional form.
- The simplified workflow form:
   - Drops tool and subworkflow steps from rendering all together and instead just renders the inputs. These are not rendered as separate "steps" or forms - but as just different inputs the same clean, simple form (more the like tool GUI). Labels (or step indexes as a fallback) are used at the input name, step annotations are used as the input help text.
   - Simplify the workflow request made by the client - send only replacement parameters and inputs - do not serialize tool state for each module. I think this makes what we are tracking more structured and should be more performant as well. Prevents storing HDA IDs in unstructured JSON blobs in the database as well.
   - Drops history target option to simplify the UI - ``simplified_workflow_run_ui_target_history`` can be set to either 'current' (default) or 'new' to adjust this parameter Galaxy-wide for the simplified form.
   - Drops job caching option to simplify the UI - ``simplified_workflow_run_ui_job_cache`` can be set to 'off' (default) or 'on' to change this parameter Galaxy-wide for the simplified form.
   - Drops resource parameters - we've already verified there are none for simplified workflows.

What is in the PR:

## Screenshots

Original Form:

<img width="539" alt="Screen Shot 2019-12-19 at 4 11 11 PM" src="https://user-images.githubusercontent.com/216771/71209873-9f2b7000-227a-11ea-90e6-52fbaf01a855.png">

Simplified Form:


<img width="551" alt="Screen Shot 2019-12-19 at 4 10 59 PM" src="https://user-images.githubusercontent.com/216771/71209877-a3f02400-227a-11ea-87cb-dcfc0e279828.png">

Clicking to open advanced form from simplified one:

![inaction](https://user-images.githubusercontent.com/216771/71210531-15c86d80-227b-11ea-9531-ade25678d46c.gif)

nsoranzo commented:
The cleanliness and simplicity of the UI is amazing, thanks @jmchilton !
@TKlingstrom will love it!

guerler commented on client/galaxy/scripts/components/WaitButton.vue:
```diff
@@ -0,0 +1,68 @@
+<template>
+    <button
+        class="ui-button-default"
+        v-bind:class="[waiting ? 'disabled' : '', variantClass]"
```
I wonder if we should avoid this one-line if statements in the template and consistently use computed values instead. I personally lean towards computed values going forward.

martenson replied:
I also think computed vs ternary has a clear winner

jmchilton replied:
Will do, thanks!

jmchilton replied:
Rebased and included in https://github.com/galaxyproject/galaxy/pull/9138 the oldest/smallest PR that contained this. Will rebase these other PRs on top of that - but they are all going to lose their selenium testing :(.

guerler commented on client/galaxy/scripts/components/Workflow/Run/WorkflowRun.vue:
```diff
@@ -0,0 +1,162 @@
+    },
+    methods: {
+        execute() {
+            this.$refs["run-form"].execute();
```
It would be nice if we could develop a cleaner way for the interaction between the vue-component and backbone sub views. Maybe a reusable wrapper in order to avoid the over use of `$refs`.

jmchilton replied:
Thanks for the review @guerler!

From https://vuejs.org/v2/guide/components-edge-cases.html

> $refs are only populated after the component has been rendered, and they are not reactive. It is only meant as an escape hatch for direct child manipulation - you should avoid accessing $refs from within templates or computed properties.
 
This usage is neither within a template or a computed property. I'm not sure I see value in avoiding ``$refs`` here - this seems like a totally fine instance of this. Is there a reason you disagree? I'm trying to understand the concern.

Since ``run-form`` can be one of two forms - I think any property handling thing would lead to duplication in the template stuff but I did look around for ideas nonetheless...

I found this:

https://www.reddit.com/r/vuejs/comments/79stdr/how_to_avoid_refs_here/

The idea I guess there would be that we addd a property "submittingForm" to the parent I suppose (WorkflowRun in this case).  And pass submittingForm to both the child elements. I think that would be kind of yucky - especially because I think we would need to add a ``watch`` to both of the child components?

```
watch: 
  formSubmitting: function(value) {
     if(value) {
        execute();
    }
  }
```

Which seems objectively more code and duplication and subjectively less intuitive to me. The top comment in that reddit thread (only 3 points I'll admit):

> As you demonstrate, the code becomes less clear, and harder to reason about in this instance.
> This seems like a good candidate for calling a specific method on your child component this.$refs.canvas.snapshot(), and having that child component fire an event which the parent captures to get the canvas data.
> The only "reaching in" you're doing is calling an available method on your child.

Maybe a wrapper like you suggest would somehow get around this sort of imitating function calls with properties and events... but I don't see it?

guerler replied:
Thanks for the detailed explanation. It makes sense to me and is a reusable pattern. A tiny improvement would be to use a name without dash to avoid quotation marks.

jmchilton replied:
Fair - happy to make that change.

jmchilton replied:
Done - and rebased into #9147.

jmchilton committed:
Simplified workflow run form.

Builds on refactoring the workflow run form landing and orchestration into Vue (#9147).

Implements #9111 - part of https://github.com/galaxyproject/galaxy/projects/15 and outlined in http://bit.ly/simplified-workflow-ui.

What is in the PR:

- Add new Galaxy configuration option - ``simplified_workflow_run_ui`` set to ``off`` by default.
- If instead the admin has set this parameter to ``prefer``, on the workflow run page scan the workflow run request and if three conditions are met show a simplfied tool form. These conditions are:
  - No workflow "resource parameters" - pretty niche, not documented, I don't think anyone outside of Canada is using them and even them I'm not sure if they ever got to production.
  - No "open" tools (i.e. tools with disconnected runtime inputs).
  - No "replacement parameters" defined outside PJAs. I'm calling #{} inside PJA and ${} inside tool steps both "replacement parameters" because the user populates them the same way in the GUI. The API only handles them inside the context of the PJA - it seems the GUI is responsible for replacing them at runtime in the traditional form.
- The simplified workflow form:
   - Drops tool and subworkflow steps from rendering all together and instead just renders the inputs. These are not rendered as separate "steps" or forms - but as just different inputs the same clean, simple form (more the like tool GUI). Labels (or step indexes as a fallback) are used at the input name, step annotations are used as the input help text.
   - Simplify the workflow request made by the client - send only replacement parameters and inputs - do not serialize tool state for each module. I think this makes what we are tracking more structured and should be more performant as well. Prevents storing HDA IDs in unstructured JSON blobs in the database as well.
   - Drops history target option to simplify the UI - simplified_workflow_run_ui_target_history can be set to either 'current' or 'new' to adjust this parameter Galaxy-wide for the simplified form.
   - Drops job caching option to simplify the UI - simplified_workflow_run_ui_cache can be set to 'off' or 'on' to change this parameter Galaxy-wide for the simplified form.
   - Drops resource parameters - we've already verified there are none for simplified workflows.

guerler commented on client/galaxy/scripts/components/Workflow/Run/WorkflowRun.vue:
```diff
@@ -83,20 +110,49 @@ export default {
+                    // simplified tool form.
+                    if (model.hasOpenToolSteps) {
+                        console.log(
+                            "cannot render simple workflow form - one or more tools have disconnected runtime inputs"
```
These probably should be shown as error messages on the screen.

jmchilton replied:
These aren't errors though (?) - if any of these conditions are met the simple the full form is just rendered. I was thinking about adding a "lint for best practices" button in the workflow editor - that might be a better spot for user messages around these?

Although - I am noticing I don't have that simple_form=false here in this branch like the others.

dannon replied:
Agreed it doesn't make sense to bubble this to the user; there's nothing for them to act on here and it's not something we can/should pause any interaction for.

bgruening reviewed (approved):
Awesome! Thanks John!

dannon commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -2436,6 +2436,38 @@ mapping:
+          future 'force' may be added an option for Galaskio-style servers that should
+          only render simplified workflows.
+
+      simplified_workflow_run_ui_target_history:
```
This and the setting below really feel like they should be user preferences, not galaxy-wide configuration settings for the sake of simplifying a page.  For more context, this notion is probably stemming from my work separating the client from the server, but this *more* tightly couples the displayed interface with the server and its configuration.

jmchilton replied:
This project stemmed from a desire to present a completely different, simplified UI around workflows - including one that doesn't even need to present the concept of a history to users. If you wanted to deploy such an app you really wouldn't want to make this a user setting. 

I'm all for decoupling - the corresponding attributes on Vue components do have defaults. They should work stand-alone in other scenarios. So they are decoupled properly in my mind. Just because we don't want to force the developer using the components to define these doesn't mean the admins shouldn't be allowed to pick defaults. Likewise, just because it would likely be good to allow user preferences in the abstract for more typical uses doesn't mean we shouldn't allow admins to be able to control how the app is configured as a default.

Having one part of the UI depend on Galaxy's concept of a user preference tab and the backend for storing and fetching those - is in fact more coupling to my mind than this. Again, the components would work without those so it seems like a fine sort of optional coupling. We have independent ways to configure these things and the component internals don't depend on how those are being configured - that would seem to be the goal to me.

Anyway, while I agree with some of this, adding the user preferences seems like outside the scope of the initial implementation. If you insist on my adding setting I will consider it to get this done, but even in that scenario I still think the defaults should be configurable by the deployer. Though I'm totally open to there being a different way for the deployer to configure the client than Galaxy's YAML.

dannon replied:
>  If you insist 

@jmchilton I really wasn't insisting on anything, sorry if you thought as much.

I think this component should be seen as a simple, togglable level of detail view rather than a different, server-set preference.  In my mind the second pass, the actual workflow-centric interface addition, that's what would add these extra properties (however they end up) and pass them to this view.  This is nice progress as a straightforward level of detail toggle on the existing workflow interface.  That's what I was trying to get at above, with the configuration defaults -- it doesn't serve anyone to force one view over the other with just this piece, I don't think, and I suppose a user could just flip the interface one way or another if they wanted anyway by swapping a bit of client-side data.

Anyway, for now, it's fair enough that until we have a way to configure more (and default) client(s), that this can live in the core galaxy configuration.  If you want this configurable at this point, then it's fine.

dannon commented on client/src/components/Workflow/Run/WorkflowRunFormSimple.vue:
```diff
@@ -0,0 +1,117 @@
+            const inputs = [];
+
+            // Add workflow parameters.
+            _.each(this.model.wpInputs, (input, i) => {
```
We should be avoiding underscore when possible at this point.

dannon commented on client/src/components/Workflow/Run/WorkflowRunFormSimple.vue:
```diff
@@ -0,0 +1,117 @@
+            _.each(this.model.wpInputs, (input, i) => {
+                const inputCopy = Object.assign({}, input);
+                // do we want to keep the color if we're not showing steps?
+                inputCopy["color"] = undefined;
```
We generally prefer dot notation.

dannon commented on client/src/components/Workflow/Run/WorkflowRunFormSimple.vue:
```diff
@@ -0,0 +1,117 @@
+        this.form = new Form(this.formDefinition());
+        this.$nextTick(() => {
+            const el = this.$refs["form"];
+            $(el).append(this.form.$el);
```
We should avoid jquery in vue components.
